### PR TITLE
Unsaved changes prompt

### PR DESCRIPTION
**Pull request by srdjan-harness**
Unsaved changes prompt implementation (ExitConfirmContext and useExitPrompt). 

Other:
- first input focus in the Run Pipeline form
- fallback for requestIdleCallback - for Safari browser

### CODE REVIEW

File: apps/gitness/src/components/ExitConfirmDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import { ExitConfirmOptions } from '../framework/context/ExitConfirmContext'
+
+export type ExitConfirmDialogProps = ExitConfirmOptions & { open: boolean }
+
+export function ExitConfirmDialog({
+  open,
+  title,
+  subtitle,
+  cancelText,
+  confirmText,
```
Comment (abhinavrastogi-harness) [inline apps/gitness/src/components/ExitConfirmDialog.tsx:19]: these should have some default values

File: apps/gitness/src/framework/context/ExitConfirmContext.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import React, { createContext, useCallback, useMemo, useState } from 'react'
+import { ExitConfirmDialog } from '../../components/ExitConfirmDialog'
+
+export interface ExitConfirmOptions {
+  title?: string
+  subtitle?: string
+  confirmText?: string
+  cancelText?: string
+  onConfirm?: () => void
+  onCancel?: () => void
```
Comment (abhinavrastogi-harness) [inline apps/gitness/src/framework/context/ExitConfirmContext.tsx:10]: I think these callbacks should be required. the display strings can have default values to make them optional

File: apps/gitness/src/framework/hooks/useExitPrompt.tsx [hunk 3]
```diff
@@ -0,0 +1,69 @@
+import { useCallback, useEffect } from 'react'
```
Comment (abhinavrastogi-harness) [inline apps/gitness/src/framework/hooks/useExitPrompt.tsx:1]: this can be a `.ts` file

File: apps/gitness/src/framework/hooks/useExitPrompt.tsx [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+  confirmText?: string
+  cancelText?: string
+}
+
+export const useExitPrompt = ({
+  isDirty = false,
+  title = 'You have unsaved changes',
+  subtitle = 'Are you sure you want to leave this page without saving?',
+  confirmText = 'Leave',
+  cancelText = 'Stay'
```
Comment (abhinavrastogi-harness) [inline apps/gitness/src/framework/hooks/useExitPrompt.tsx:17]: these defaults should be on the component I think.. not on the hook

File: apps/gitness/src/pages/run-pipeline-dialog/run-pipeline-form.tsx [hunk 5]
```diff
@@ -112,6 +114,16 @@ export default function RunPipelineForm({
         // TODO: error toast here ?
       })
   }
+
+  const formRef = useRef<HTMLDivElement | null>(null)
+  // NOTE: custom focus implementation
```
Comment (abhinavrastogi-harness) [inline apps/gitness/src/pages/run-pipeline-dialog/run-pipeline-form.tsx:119]: why do we need a custom focus implementation?
can we just set `autofocus` attribute on the first form input instead?

**Review by abhinavrastogi-harness — APPROVED** [review]

File: apps/gitness/src/components/ExitConfirmDialog.tsx (hunk 1 above)
Comment (srdjan-harness) [reply apps/gitness/src/components/ExitConfirmDialog.tsx]: Updated. Defaults are set in component.

File: apps/gitness/src/framework/context/ExitConfirmContext.tsx (hunk 2 above)
Comment (srdjan-harness) [reply apps/gitness/src/framework/context/ExitConfirmContext.tsx]: Updated. Callbacks are mandatory.

File: apps/gitness/src/pages/run-pipeline-dialog/run-pipeline-form.tsx (hunk 5 above)
Comment (srdjan-harness) [reply apps/gitness/src/pages/run-pipeline-dialog/run-pipeline-form.tsx]: Fixed. We are using setFocus that is exposed via useForm in the form library. Added autoFocusPath to RootForm.

**Pull request merged**